imqsauth/audit.go

```diff
@@ -0,0 +1,67 @@
+package imqsauth
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/IMQS/log"
+	"github.com/IMQS/serviceauth"
+)
+
+type IMQSAuditor struct {
+	Url string
+	Log *log.Logger
+}
+
+func NewIMQSAuditor(url string, logger *log.Logger) *IMQSAuditor {
+	return &IMQSAuditor{Url: url, Log: logger}
+}
+
+type Action struct {
+	Who     string        `json:"who"`
+	DidWhat string        `json:"did_what"`
+	ToWhat  string        `json:"to_what"`
+	AtTime  int64         `json:"at_time"`
+	Context ActionContext `json:"context"`
+}
+
+type ActionContext struct {
+	Location string `json:"location"`
+}
+
+func (a *IMQSAuditor) AuditUserAction(identity, clientIp, description string) {
+	action := &Action{Who: identity, DidWhat: description, AtTime: time.Now().Unix()}
+
+	var actionContext ActionContext
```

Comment:
No need to create an `ActionContext`.
You can just assign to member directly, i.e. `action.Context.Location = clientIp`